### ISSUE TITLE
Pagination -- subscription not updating within $meteor.autorun()

### ISSUE THREAD

**Issue by MilosStanic** [Urigo/angular-meteor#887]
I made an app with almost identical example to angular-meteor-socially for pagination. 
However, it doesn't work. 
When I clone angular-meteor-socially and run it on my local machine it works
When I copy the search-pagination example code from socially to my app, it doesn't work. 

Here's the relevant snippets: 

```javascript
angular.module('admin').controller('AdminCtrl', function($scope, $meteor, $stateParams, FileUpload){
	//var vm = this;

	$scope.page = 1;
  	$scope.perPage = 3;
  	$scope.sort = {username: 1};
  	$scope.orderProperty = '1';
  	
	$scope.chatUsers = $scope.$meteorCollection(function() {
		return Meteor.users.find({}, {			
			sort : $scope.getReactively('sort')
		});
	});	
	
	$meteor.autorun($scope, function() {
	    $meteor.subscribe('myusers', {
			limit: parseInt($scope.getReactively('perPage')),
			skip: (parseInt($scope.getReactively('page')) - 1) * parseInt($scope.getReactively('perPage')),
			sort: $scope.getReactively('sort')
	    }, $scope.getReactively('search')).then(function(){	    	
			$scope.usersCount = $meteor.object(Counts ,'numberOfUsers', false);
			//$scope.usersCount = Counts.get('numberOfUsers');
			console.log('updating users count', $scope.usersCount);
	  	});
	});			
	
	$scope.pageChanged = function(newPage) {
		$scope.page = newPage;
	};
```

And here's the server part: 

```javascript
Meteor.publish('myusers', function (options, searchString) {
	if (searchString == null)
    	searchString = '';
	console.log('getting options', options, 'searchString', searchString);
	Counts.publish(this, 'numberOfUsers', Meteor.users.find({'username' : { '$regex' : '.*' + searchString || '' + '.*', '$options' : 'i' }}, 
		{fields: {username: 1, emails: 1, profile: 1}}), { noReady: true });

	return Meteor.users.find({'username' : { '$regex' : '.*' + searchString || '' + '.*', '$options' : 'i'}}, 
		{fields: {username: 1, emails: 1, profile: 1}}, options);	
});
```

You can see the whole app in a repo: 
https://bitbucket.org/milosstanic/chatapp/src

I've been at it for two whole days, and I've ran out of ideas.

Thanks.

**Comment by MilosStanic** [comment]
btw, the app won't run unless you first clone a small package to your /packages dir
so cd to /packages and run 
`git clone https://github.com/rclai/meteor-lodash-replace-underscore.git`
Then create an user, and make it admin by editing the /server/roles.js and replacing the two _id fields with your newly created user's _id.
Then when you login as admin, you'll see ADMIN in top bar on the right. Click on it, and then click on the hamburger on the left in the top bar. A sidebar should open and then click users list. And there you go, a list that doesn't work.

**Comment by MilosStanic** [comment]
If I change the code by stopping subscription on each autorun, I get for a moment a glimpse of correct results, but then subscription resets to only the logged in user. 

Here's how I stop subscription: 

```
...
}, $scope.getReactively('search')).then(function(subhandle){         
            $scope.usersCount = $meteor.object(Counts ,'numberOfUsers', false);
            //$scope.usersCount = Counts.get('numberOfUsers');
            console.log('updating users count', $scope.usersCount);
            subhandle.stop();
        });
...
```
Any ideas? Please help!!!

**Comment by MilosStanic** [comment]
I figured out, using `msavin:mongol` that when I turn off all subscriptions in the controller, I'm still subscribed to `users` and the subscription contains only one document -- the currently logged in user. I can't find that I ever subscribed to it. Is it something angular-meteor does automatically? I'm betting this subscription is messing with my pagination and autorun.

**Comment by MilosStanic** [comment]
Don't know what I did... I removed some other things from my controller, and now the search filter works, but pagination doesn't work yet. 
Will see how things turn out.

**Comment by MilosStanic** [comment]
There's definitely some lurking subscription to `users` and it kills the joy of pagination and search. 
Is it possible that listing of users can't be done due to some internal restriction in angular-meteor? @Urigo please help!

**Comment by Urigo** [comment]
angular-meteor doesn't subscribe to anything without you explicitly asking it..
maybe you have an unclosed subscription on another controller?

**Comment by MilosStanic** [comment]
Dear @Urigo , et al.

I have solved the main problem. It was stupid as usual. I sent an extra parameter to limit fileds returned from document to `Meteor.users.find()` together and before the `options` parameter, coming from Angular. It was the `{fields: {username: 1, emails: 1, profile: 1}}`. So I removed that from params on server, and put it into subscription function on client, together with `skip`, and `limit` params, and now it works. 

However, another **problem** remains!

I'm seeing a subscription to `users` made on login. And the only document in that subscription is the current logged in user data. 
I'm using `accounts` module by @Tallyb from her angular-meteor-material boilerplate app here https://github.com/Tallyb/angular-meteor-material

But I don't see any subscriptions made to `users` in her module.
Users are logged in using plain old `$meteor.loginWithPassword` , nothing fancy there. Yet, a subscription to `users` is made, and it screws my search and user details view, since I can't kill it. 

In all search results collections, the document with current user data is added, regardless of the search parameter!

I also made a controller to view user details, here it is: 

```javascript
angular.module('admin').controller('AdminUserDetailsCtrl', function($scope, $meteor, $stateParams){
	var vm = this;
	
	vm.currentUser = $meteor.object(Meteor.users, $stateParams.userId);
	//vm.currentUser.stop();
});
```
Unless I stop the subscription as in the last line of the controller -- meteor keeps firing subscribe/unsubscribe events (I'm not sure, but msavin:mongol shows `users` collection flickering and having 1/2 results, of which, one is the ghostly logged in user, and the second one is an object/document containing only the required user's ID), and my result is empty. 

If I stop the subscription (uncomment the last line of controller), the result displays fine. Then the `users` collection contains only the ghostly logged in user data, and nothing else. I'm guessing msavin:mongol is not showing `$meteor.object` subscriptions.  

This is a really really strange situation. 

Thanks for your time and patience.

**Comment by mxab** [comment]
@MilosStanic calling stop on an meteor object has nothing todo with the subscription http://www.angular-meteor.com/api/AngularMeteorObject

**Comment by mxab** [comment]
Just as a side note, I think it is actually not possible to model a correct paging mechanism with meteor publish/subscribe as long as you cannot enforce only a single subscription of collection. For a correct paging you might should consider a more classical "request a page via a method call"

**Comment by MilosStanic** [comment]
@mxab okay. I know it hasn't anything to do with a subscription, however, without stopping the meteor object, it doesn't work. And I explained how it behaves. I wish someone could explain why this happens. 

Pagination works as it is. And it works fine, so far. At least as far as I can see.

Feel free to look at my source code at the link from the first post. 

Thank you very much.

**Comment by Tallyb** [comment]
@MilosStanic  any luck?

**Comment by MilosStanic** [comment]
No, I've given up... I'll do as @mxab says, I'll do some alternative pagination methods. I mean, the pagination works, it's just I'm totally unable to filter out the current user from each page. I could hack it, but that's not a solution.
However, now I'm facing some problems with uploading images and getting their urls immediately after upload. I'll open a new issue about that.

**Comment by jmwierzbicki** [comment]
I have issue with angular-meteor, i'm trying to make use of $scope.count = $meteor.object(Counts, 'numberOfItems', false) with autorun, this object seems to be reactive, because i can see how its updating with {{count.count}} however i would like to do some logic whenever this value change, and i literaly can't force it to init autorun - Can anyone tell me how to achieve such task?

**Comment by MilosStanic** [comment]
@jmwierzbicki 
I'm not sure I understood, but you can watch for changes of `$scope.count` with `$scope.$watch` like this: 
```javascript
$scope.col = $meteor.object(Counts, 'numberOfAspectLists', true); 
$scope.$watch('col', function(newValue, oldValue){
    ... // do something here 
});
```

**Comment by Urigo** [comment]
Well, maybe now with the new [Angular Meteor 1.3 release](http://info.meteor.com/blog/angular-meteor-1.3) it would be easier as you can use the regular Meteor API without wrappers.
Check out the new pagination chapter updated to 1.3 - http://www.angular-meteor.com/tutorials/socially/angular1/search-sort-pagination-and-reactive-vars and let me know if the issue is still an Angular Meteor issue or a Meteor issue

**Comment by MilosStanic** [comment]
I just reimplemented the thing in angular-meteor 1.3, and the results are mixed. 

It does work better, it draws limited user sets, based on perPage number. 

However, the reactive searchText is not really reactive, I don't see it sending the search parameter to server to refine dataset. Maybe I have some bug, I'll keep at it and report. 

The issue with the ghost current user appearing on each page is somewhat better: now it doesn't appear on all pages, only on some :) . 

Namely, the current username begins with 'D', so in paginated results, it doesn't appear on first three pages, but as soon as the paginator hits the page where this user appears, it appears on all pages after that. In my example, it doesn't appear on first three pages, but only after that.  :)

**Comment by Urigo** [comment]
@MilosStanic maybe that means it's not an Angular specific problem?

**Comment by Deadly0** [comment]
@Urigo this is not a bug. 
@MilosStanic to solve your problem you need:

1) Create client only collection (call it 'userList'). To do it place this code in /client folder (i prefer in /client/lib/collections/user-list.js).
```
var UserList = new Mongo.Collection('userList');
```

2) Make custom publish method. 
```
Meteor.publish('userList', function (query, opts /* Any params you like */) {
  // Any logic here to prepear the query
  var self = this;
  Meteor.users.find(query, opts).forEach(function(user)  {
     //               v The name of our collection
     self.added('userList', user._id, user);
  });

  this.ready();
});
```

2) Now you are ready to use the data.
```
$scope.chatUsers = $scope.$meteorCollection(function() {
        return UserList.find({}, {          
            sort : $scope.getReactively('sort')
        });
    }); 

// ....

$meteor.subscribe('userList', { //.....
```
All other code should be the same.

**Comment by MilosStanic** [comment]
Thank you! I will try this and report back results.

**Comment by MilosStanic** [comment]
@Deadly0 **THANK YOU VERY VERY MUCH!!!!**

You totally saved me with this. 

However, there's one tiny error you have: in point 1. there shouldn't be `var`, so just: 
`UserList = new Mongo.Collection('userList');`

Regarding pagination, I only have to add that the publish counts package works normally. 
Reading on the docs for Meteor.publish http://docs.meteor.com/#/full/meteor_publish led me to how to adapt your code to mine. 

Everything fine for now, as far as I can see.

Thank you again!
